superset/migrations/versions/2023-03-27_12-30_7e67aecbf3f1_chart_ds_constraint.py

```diff
@@ -0,0 +1,66 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""chart-ds-constraint
+
+Revision ID: 7e67aecbf3f1
+Revises: b5ea9d343307
+Create Date: 2023-03-27 12:30:01.164594
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = "7e67aecbf3f1"
+down_revision = "07f9a902af1b"
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.ext.declarative import declarative_base
+
+from superset import db
+
+Base = declarative_base()
+
+
+class Slice(Base):  # type: ignore
+    __tablename__ = "slices"
+
+    id = sa.Column(sa.Integer, primary_key=True)
+    slice_name = sa.Column(sa.String(250))
+    datasource_type = sa.Column(sa.String(200))
+    query_context = sa.Column(sa.Text)
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    with op.batch_alter_table("slices") as batch_op:
+        for slc in session.query(Slice).filter(Slice.datasource_type != "table").all():
+            # clean up all charts with datasource_type not != table
+            slc.datasource_type = "table"
```

Comment:
@hughhhh we're seeing a defined `slices.datasource_name` column when `slices.datasource_type = 'query'` however it doesn't seem to be a dataset in the `tables` table—which would indicate that your migration would likely break said slices when you make the switch from `query` to `table`. BTW where is this query stored?

Reply:
> @hughhhh we're seeing a defined `slices.datasource_name` column when `slices.datasource_type = 'query'` however it doesn't seem to be a dataset in the `tables` table—which would indicate that your migration would likely break said slices when you make the switch from `query` to `table`. BTW where is this query stored?

is the column `datasource_name` only appearing when the `datasource_type` is `query`?

Query is defined [here](https://github.com/apache/superset/blob/1dd895b9b0ebc2efcd2193b72013b4f9f0183564/superset/models/sql_lab.py#L65)

Reply:
this is the stack trace we are encountering in production:
```
return [
  File "/usr/local/lib/python3.8/site-packages/marshmallow/schema.py", line 515, in <listcomp>
    self._serialize(d, many=False)
  File "/usr/local/lib/python3.8/site-packages/marshmallow/schema.py", line 520, in _serialize
    value = field_obj.serialize(attr_name, obj, accessor=self.get_attribute)
  File "/usr/local/lib/python3.8/site-packages/marshmallow/fields.py", line 338, in serialize
    return self._serialize(value, attr, obj, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/marshmallow/fields.py", line 1922, in _serialize
    return self._call_or_raise(self.serialize_func, obj, attr)
  File "/usr/local/lib/python3.8/site-packages/marshmallow/fields.py", line 1936, in _call_or_raise
    return func(value)
  File "/usr/local/lib/python3.8/site-packages/superset/models/slice.py", line 167, in datasource_url
    return datasource.explore_url if datasource else None
AttributeError: 'Query' object has no attribute 'explore_url'
```

Reply:
@hughhhh per your response,

> So those slices are most likely broken, because charts/slices don't know who to execute grabbing data when the datasource_type is Query

I guess the question/concern is simply changing the `datasource_type` to `table` doesn't really mitigate the problematic charts as the `datasource_id` is then likely incorrect as well, i.e., it's referencing the `query.id` column rather than the `tables.id` column.

Also please refer to my comment regarding the chart parameters as we (for right or wrong) doubly define the ID/type tuple.

Reply:
so the idea here was to try and see if some of them will work if we change the type, since we can't just delete these charts. We are assuming these charts are failing already and hoping this might help fix it, but no guarantee. Also now with the constraint we should see error when the user/api tries to write slice.datasource_type = query